edx_proctoring/backends/software_secure.py

```diff
@@ -0,0 +1,229 @@
+"""
+
+from Crypto.Cipher import DES3
+import base64
+from hashlib import sha256
+import requests
+import hmac
+import binascii
+import datetime
+import json
+import logging
+
+from edx_proctoring.backends.backend import ProctoringBackendProvider
+from edx_proctoring.exceptions import BackendProvideCannotRegisterAttempt
+
+
+log = logging.getLogger(__name__)
+
+
+class SoftwareSecureBackendProvider(ProctoringBackendProvider):
+    """
+    Implementation of the ProctoringBackendProvider for Software Secure's
+    RPNow product
+    """
+
+    def __init__(self, organization, exam_sponsor, exam_register_endpoint,
+                 secret_key_id, secret_key, crypto_key, software_download_url):
+        """
+        Class initializer
+        """
+
+        self.organization = organization
+        self.exam_sponsor = exam_sponsor
+        self.exam_register_endpoint = exam_register_endpoint
+        self.secret_key_id = secret_key_id
+        self.secret_key = secret_key
+        self.crypto_key = crypto_key
+        self.timeout = 10
+        self.software_download_url = software_download_url
+
+    def register_exam_attempt(self, exam, time_limit_mins, attempt_code,
+                              is_sample_attempt, callback_url):
+        """
+        Method that is responsible for communicating with the backend provider
+        to establish a new proctored exam
+        """
+
+        data = self._get_payload(
+            exam,
+            time_limit_mins,
+            attempt_code,
+            is_sample_attempt,
+            callback_url
+        )
+        headers = {
+            "Content-Type": 'application/json'
+        }
+        http_date = datetime.datetime.utcnow().strftime("%a, %d %b %Y %H:%M:%S GMT")
+        signature = self._sign_doc(data, 'POST', headers, http_date)
+
+        status, response = self._send_request_to_ssi(data, signature, http_date)
+
+        if status not in [200, 201]:
+            err_msg = (
+                'Could not register attempt_code = {attempt_code}. '
+                'HTTP Status code was {status_code} and response was {response}.'.format(
+                    attempt_code=attempt_code,
+                    status_code=status,
+                    response=response
+                )
+            )
+            raise BackendProvideCannotRegisterAttempt(err_msg)
+
+        # get the external ID that Software Secure has defined
+        # for this attempt
+        ssi_record_locator = json.loads(response)['ssiRecordLocator']
+
+        return ssi_record_locator
+
+    def start_exam_attempt(self, exam, attempt):  # pylint: disable=unused-argument
+        """
+        Called when the exam attempt has been created but not started
+        """
+        return None
+
+    def stop_exam_attempt(self, exam, attempt):
+        """
+        Method that is responsible for communicating with the backend provider
+        to establish a new proctored exam
+        """
+        return None
```

Comment:
As with above, returning None here seems odd -- i guess I would have expected these operations to do something.